### DISCUSSION TITLE
Issue in getInitialProps after upgrading NextJs to v9.5.3

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#17217] [Help]
Previously I was using the NextJs v9.4.4, in that version getInitialProps was working fine but after upgrading it to the latest version I was getting an error which says: context.req is undefined.
Here is the code snippet in which I was getting error: 
```
LessonPage.getInitialProps = async context => {
  const { whiteboardEvents, lesson } = context.req;
  return {
    whiteboardEvents,
    lesson,
  };
};
```

**Comment by jamesmosier** [comment]
Hi @siddhantj007, do you have a [custom `_app.js`](https://nextjs.org/docs/advanced-features/custom-app)? 

If so, are you passing `context` to `App.getInitialProps`?

```js
MyApp.getInitialProps = async (appContext) => {
  // calls page's `getInitialProps` and fills `appProps.pageProps`
  const appProps = await App.getInitialProps(appContext);

  return { ...appProps }
}
```

---

*Edit*:

It should be:

```js
export async function getServerSideProps(context) {
  const { whiteboardEvents, lesson } = context.query;
  return {
    props: {
      whiteboardEvents: whiteboardEvents || null,
      lesson: parse(stringify(lesson)),
    },
  };
}
```

`context.query` is how you get the query string. [docs](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering)

---

```js
LessonPage.getInitialProps = async context => {
  const { whiteboardEvents, lesson } = context.query;
  return {
    whiteboardEvents,
    lesson,
  };
};
```

[docs for getInitialProps](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps#context-object)

**Reply by ghost** [reply]
Hi @jamesmosier , thanks for the reply. 
Yes, I do have custom `_app.js`, below is the snippet of that.

```
MyApp.getInitialProps = async ({ Component, ctx }) => {
  ctx.store.dispatch({ type: 'TOE', payload: 'was set in _app' });
 return {
    pageProps: {
      // Call page-level getInitialProps
      ...(Component.getInitialProps
        ? await Component.getInitialProps(ctx)
        : {}),
      // Some custom thing for all pages
      pathname: ctx.pathname,
    },
    store: { getState: () => ctx.store.getState },
  };
};
```

**Reply by kartikag01** [reply]
Try this, it should solve the issue
```
  MyApp.getInitialProps = async ({ Component, ctx }) => {
  ctx.store.dispatch({ type: "TOE", payload: "was set in _app" });
  let pageProps = {};

  if (Component.getInitialProps) {
    pageProps = await Component.getInitialProps(ctx);
  }

  // this modification is required
  if (Object.keys(pageProps).length > 0) {
    return { pageProps, store: { getState: () => ctx.store.getState } };
  }
  return { store: { getState: () => ctx.store.getState } };
  // return {};
};

```

**Reply by ghost** [reply]
@kartikag01 where should I exactly put this. My `_app.js` is a functional component. When I try to write this outside the component I get the error in the first line, to be specific, error is on `static` and `getInitialProps`. If I write this in the component then also the same error.

**Reply by kartikag01** [reply]
@siddhantj007 i have updated the method, pls check that.

**Reply by ghost** [reply]
@kartikag01 I tried this, but this didn't work out. It still gives the same error i.e `context.req` is undefined.

**Reply by ghost** [reply]
> Try this, it should solve the issue
> 
> ```
>   MyApp.getInitialProps = async ({ Component, ctx }) => {
>   ctx.store.dispatch({ type: "TOE", payload: "was set in _app" });
>   let pageProps = {};
> 
>   if (Component.getInitialProps) {
>     pageProps = await Component.getInitialProps(ctx);
>   }
> 
>   // this modification is required
>   if (Object.keys(pageProps).length > 0) {
>     return { pageProps, store: { getState: () => ctx.store.getState } };
>   }
>   return { store: { getState: () => ctx.store.getState } };
>   // return {};
> };
> ```
@kartikag01 Why did you remove `pathname` from `pageProps`?

**Reply by kartikag01** [reply]
oho, i missed that.
the basic idea is to return `{}` when  `Object.keys(pageProps).length` is zero instead of  `{ pageProps }`

**Reply by ghost** [reply]
> Previously I was using the NextJs v9.4.4, in that version getInitialProps was working fine but after upgrading it to the latest version I was getting an error which says: context.req is undefined.
> Here is the code snippet in which I was getting error:
> 
> ```
> LessonPage.getInitialProps = async context => {
>   const { whiteboardEvents, lesson } = context.req;
>   return {
>     whiteboardEvents,
>     lesson,
>   };
> };
> ```
@jamesmosier @kartikag01 I have just replace the getInitialProps with getServerSideProps, and this works in dev environment but in production, the error still remains, any idea why is that?

**Reply by jamesmosier** [reply]
Unfortunately I'm not certain without seeing more code.

**Reply by ghost** [reply]
@jamesmosier This is how I implemented getServerSideProps.
```
const LessonPage = ({ whiteboardEvents, lesson }) => {
  return (
    <>
      {typeof window !== 'undefined' && (
        <Head>
          <title>Vidyalai Collaborative Classroom Beta</title>
        </Head>
      )}
      <LessonMeeting lesson={lesson} whiteboardEvents={whiteboardEvents} />
    </>
  );
};

export async function getServerSideProps(context) {
  const { whiteboardEvents, lesson } = context.req;
  return {
    props: {
      whiteboardEvents: whiteboardEvents || null,
      lesson: parse(stringify(lesson)),
    },
  };
}
```

**Reply by jamesmosier** [reply]
It should be:

```js
export async function getServerSideProps(context) {
  const { whiteboardEvents, lesson } = context.query;
  return {
    props: {
      whiteboardEvents: whiteboardEvents || null,
      lesson: parse(stringify(lesson)),
    },
  };
}
```

`context.query` is how you get the query string. [docs](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering)

---

```js
LessonPage.getInitialProps = async context => {
  const { whiteboardEvents, lesson } = context.query;
  return {
    whiteboardEvents,
    lesson,
  };
};
```

[docs for getInitialProps](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps#context-object)

**Reply by ghost** [reply]
@jamesmosier By replacing `req` with `query`, I get status code 500.

**Reply by ghost** [reply]
```
            <Link
              href={
                lesson.isDemo
                  ? `/classrooms/room/demo`
                  : `/lessons/${lesson.meetingId}`
              }
            >
              <Button variant={'contained'} className={'joinLesson'}>
                <span className={'text'}>JOIN LESSON</span>
                <Icon>keyboard_arrow_right</Icon>
              </Button>
            </Link>
```
@jamesmosier Is it necessary to include an anchor tag in the above code and if yes then how?

**Reply by ghost** [reply]
@jamesmosier Issue is solved now, the problem was with the API call.
